### ISSUE TITLE
PyTorch distributed training hangs on broadcast_parameters()

### ISSUE THREAD

**Issue by wangli1426** [bytedance/byteps#267]
**Describe the bug**
I am trying to run PyTorch distributed training on 2 GPU nodes, following the guidance on https://github.com/bytedance/byteps/blob/master/docs/step-by-step-tutorial.md#distributed-training-tcp.  However, both workers hang on `bps.broadcast_parameters(model.state_dict(), root_rank=0)`.

**To Reproduce**
1. Start scheduler on node1.
```bash
docker run -it --net=host bytepsimage/pytorch bash
export DMLC_NUM_WORKER=2
export DMLC_ROLE=scheduler
export DMLC_NUM_SERVER=1
export DMLC_PS_ROOT_URI=0.0.0.0 # the scheduler IP
export DMLC_PS_ROOT_PORT=12345  # the scheduler port
bpslaunch
```
I saw the following logs:
```
BytePS launching scheduler
[03:59:10] byteps/server/server.cc:339: BytePS server engine uses 4 threads, consider increasing BYTEPS_SERVER_ENGINE_THREAD for higher performance
[03:59:10] src/./zmq_van.h:287: Start ZMQ recv thread
```
2. Start worker on node2
```bash
docker run -it --net=host bytepsimage/pytorch bash
export DMLC_NUM_WORKER=2
export DMLC_ROLE=server
export DMLC_NUM_SERVER=1
export DMLC_PS_ROOT_URI=node1 # the scheduler IP
export DMLC_PS_ROOT_PORT=12345  # the scheduler port
bpslaunch
```
I saw the log:
```
BytePS launching server
[03:59:12] byteps/server/server.cc:339: BytePS server engine uses 4 threads, consider increasing BYTEPS_SERVER_ENGINE_THREAD for higher performance
[03:59:12] src/./zmq_van.h:287: Start ZMQ recv thread
```
3. Start trainer on node1
```
export NVIDIA_VISIBLE_DEVICES=0,1,2,3
export DMLC_WORKER_ID=0
export DMLC_NUM_WORKER=2
export DMLC_ROLE=worker
export DMLC_NUM_SERVER=1
export DMLC_PS_ROOT_URI=node1 # the scheduler IP
export DMLC_PS_ROOT_PORT=12345 # the scheduler port
bpslaunch python3 /usr/local/byteps/example/pytorch/benchmark_byteps.py --model resnet50 --num-iters 1000000
```
The logs:
```
BytePS launching worker
parse...
parse...
parse...
parse...
init...
init...
init...
init...
optim...
optim...
broadcast_parameters...
[04:01:18] src/./zmq_van.h:287: Start ZMQ recv thread
[04:01:18] src/./zmq_van.h:287: Start ZMQ recv thread
[04:01:18] src/./zmq_van.h:287: Start ZMQ recv thread
optim...
broadcast_parameters...
optim...
broadcast_parameters...
broadcast_parameters...
```
4. Start trainer on node2
```
export NVIDIA_VISIBLE_DEVICES=0,1,2,3
export DMLC_WORKER_ID=1
export DMLC_NUM_WORKER=2
export DMLC_ROLE=worker
export DMLC_NUM_SERVER=1
export DMLC_PS_ROOT_URI=node1 # the scheduler IP
export DMLC_PS_ROOT_PORT=12345 # the scheduler port
bpslaunch python3 /usr/local/byteps/example/pytorch/benchmark_byteps.py --model resnet50 --num-iters 1000000
```
The logs:
```
BytePS launching worker
parse...
init...
parse...
init...
parse...
init...
parse...
init...
optim...
optim...
optim...
optim...
broadcast_parameters...
[04:01:18] src/./zmq_van.h:287: Start ZMQ recv thread
broadcast_parameters...
broadcast_parameters...
broadcast_parameters...
[04:01:18] src/./zmq_van.h:287: Start ZMQ recv thread
[04:01:18] src/./zmq_van.h:287: Start ZMQ recv thread
```

Note I modified file `example/pytorch/benchmark_byteps.py` as following to print the log:
```python
...
log_and_flush("optim...")
optimizer = optim.SGD(model.parameters(), lr=0.01)

# BytePS: (optional) compression algorithm.
compression = bps.Compression.fp16 if args.fp16_pushpull else bps.Compression.none

# BytePS: wrap optimizer with DistributedOptimizer.
optimizer = bps.DistributedOptimizer(optimizer,
                                     named_parameters=model.named_parameters(),
                                     compression=compression)

# BytePS: broadcast parameters & optimizer state.
log_and_flush("broadcast_parameters...")
bps.broadcast_parameters(model.state_dict(), root_rank=0)
log_and_flush("broadcast_optimizer_state...")
bps.broadcast_optimizer_state(optimizer, root_rank=0)
...
```

**Expected behavior**
The distributed training should start to train.

**Environment (please complete the following information):**
 - OS: Ubuntu 16.04(host)
 - GCC version: gcc version 7.4.0 
 - CUDA and NCCL version: See details in additional context
 - Framework (TF, PyTorch, MXNet): PyTorch

**Additional context**
I got the following version information in the container. 
```bash
>>> import torch
>>> torch.cuda.nccl.version()
2408
>>> torch.version.cuda
'10.1'
>>> torch.cuda.nccl.version()
2408
>>> torch.backends.cudnn.version()
7603
```

The hash of the docker image is:
```bash
docker pull bytepsimage/pytorch
Using default tag: latest
latest: Pulling from bytepsimage/pytorch
Digest: sha256:7b5db6b34706d8b10e2cbc3dda9fd362e42dc98dceb21d6eb90ef060bd76a5ad
Status: Image is up to date for bytepsimage/pytorch:latest
```

**Comment by bobzhuyb** [comment]
It seems to me that neither the server or workers can connect to your scheduler. After you start the scheduler and server, can you run `netstat -ant |grep 12345` and check whether the server is connected to the scheduler? 

Also, can you set `PS_VERBOSE=2` and update the logs you see?

**Comment by wangli1426** [comment]
@bobzhuyb Thanks for the prompt rely.

Following your suggestion, I checked the TCP connection with `netstat` command and I found the connection between the server and the scheduler had been established.:
```
# netstat -anp|grep 12345
[11:16:09] src/./zmq_van.h:66: BYTEPS_ZMQ_NTHREADS set to 4                      │tcp        0      0 10.80.40.12:35184       10.80.40.11:12345        ESTABLISHED 
```

Also, by setting ``PS_VERBOSE=2``, the scheduler printed more logs, which indicated that the workers and the server were connected properly?  
```
BytePS launching scheduler
[11:15:26] byteps/server/server.cc:339: BytePS server engine uses 4 threads, consider increasing BYTEPS_SERVER_ENGINE_THREAD for higher performance
[11:15:26] src/./zmq_van.h:61: BYTEPS_ZMQ_MAX_SOCKET set to 1024
[11:15:26] src/./zmq_van.h:66: BYTEPS_ZMQ_NTHREADS set to 4
[11:15:26] src/van.cc:389: Bind to [role=scheduler, id=1, ip=0.0.0.0, port=12345, is_recovery=0
11:15:26] src/./zmq_van.h:287: Start ZMQ recv thread
[11:15:43] src/van.cc:503: ? => 1. Meta: request=0, timestamp=0, control={ cmd=ADD_NODE, node={ role=server, ip=10.80.40.12, port=39533, is_recovery=0 } }. THIS IS NOT DATA MSG!
[11:16:09] src/van.cc:503: ? => 1. Meta: request=0, timestamp=0, control={ cmd=ADD_NODE, node={ role=worker, ip=10.80.40.11, port=54181, is_recovery=0 } }. THIS IS NOT DATA MSG!
[11:16:19] src/van.cc:503: ? => 1. Meta: request=0, timestamp=0, control={ cmd=ADD_NODE, node={ role=worker, ip=10.80.40.12, port=52439, is_recovery=0 } }. THIS IS NOT DATA MSG!
[11:16:19] src/van.cc:140: assign rank=8 to node role=server, ip=10.80.40.12, port=39533, is_recovery=0
[11:16:19] src/van.cc:140: assign rank=9 to node role=worker, ip=10.80.40.12, port=52439, is_recovery=0
[11:16:19] src/van.cc:140: assign rank=11 to node role=worker, ip=10.80.40.11, port=54181, is_recovery=0
[11:16:19] src/van.cc:478: ? => 9. Meta: request=0, timestamp=0, control={ cmd=ADD_NODE, node={ role=server, id=8, ip=10.80.40.12, port=39533, is_recovery=0 role=worker, id=9, ip=10.80.40.12, port=52439, is_recovery=0 role=worker, id=11, ip=10.80.40.11, port=54181, is_recovery=0 role=scheduler, id=1, ip=0.0.0.0, port=12345, is_recovery=0 } }. THIS IS NOT DATA MSG!
[11:16:19] src/van.cc:478: ? => 11. Meta: request=0, timestamp=1, control={ cmd=ADD_NODE, node={ role=server, id=8, ip=10.80.40.12, port=39533, is_recovery=0 role=worker, id=9, ip=10.80.40.12, port=52439, is_recovery=0 role=worker, id=11, ip=10.80.40.11, port=54181, is_recovery=0 role=scheduler, id=1, ip=0.0.0.0, port=12345, is_recovery=0 } }. THIS IS NOT DATA MSG!
[11:16:19] src/van.cc:478: ? => 8. Meta: request=0, timestamp=2, control={ cmd=ADD_NODE, node={ role=server, id=8, ip=10.80.40.12, port=39533, is_recovery=0 role=worker, id=9, ip=10.80.40.12, port=52439, is_recovery=0 role=worker, id=11, ip=10.80.40.11, port=54181, is_recovery=0 role=scheduler, id=1, ip=0.0.0.0, port=12345, is_recovery=0 } }. THIS IS NOT DATA MSG!
[11:16:19] src/van.cc:166: the scheduler is connected to 2 workers and 1 servers
[11:16:19] src/van.cc:503: 11 => 1. Meta: request=1, timestamp=1, control={ cmd=BARRIER, barrier_group=7 }. THIS IS NOT DATA MSG!
[11:16:19] src/van.cc:265: Barrier count for 7 : 1
[11:16:19] src/van.cc:478: ? => 1. Meta: request=1, timestamp=3, control={ cmd=BARRIER, barrier_group=7 }. THIS IS NOT DATA MSG!
[11:16:19] src/van.cc:503: 1 => 1. Meta: request=1, timestamp=3, control={ cmd=BARRIER, barrier_group=7 }. THIS IS NOT DATA MSG!
[11:16:19] src/van.cc:265: Barrier count for 7 : 2
```

Any suggestion?

**Comment by bobzhuyb** [comment]
Can you try `BYTEPS_LOG_LEVEL=TRACE`?

**Comment by wangli1426** [comment]
@bobzhuyb 
I set `BYTEPS_LOG_LEVEL=TRACE` and see the following logs. I did see any error on my side. I attach the logs for your reference. Thank you.

logs on the scheduler
```
[07:15:52] byteps/server/server.cc:339: BytePS server engine uses 4 threads, consider increasing BYTEPS_SERVER_ENGINE_THREAD for higher performance
[07:15:52] src/./zmq_van.h:61: BYTEPS_ZMQ_MAX_SOCKET set to 1024
[07:15:52] src/./zmq_van.h:66: BYTEPS_ZMQ_NTHREADS set to 4
[07:15:52] src/van.cc:389: Bind to role=scheduler, id=1, ip=0.0.0.0, port=1234, is_recovery=0
[07:15:52] src/./zmq_van.h:287: Start ZMQ recv thread
[07:16:03] src/van.cc:503: ? => 1. Meta: request=0, timestamp=0, control={ cmd=ADD_NODE, node={ role=server, ip=10.80.40.12, port=56485, is_recovery=0 } }. THIS IS NOT DATA MSG!
[07:16:19] src/van.cc:503: ? => 1. Meta: request=0, timestamp=0, control={ cmd=ADD_NODE, node={ role=worker, ip=10.80.40.11, port=52213, is_recovery=0 } }. THIS IS NOT DATA MSG!
[07:16:23] src/van.cc:503: ? => 1. Meta: request=0, timestamp=0, control={ cmd=ADD_NODE, node={ role=worker, ip=10.80.40.12, port=50033, is_recovery=0 } }. THIS IS NOT DATA MSG!
[07:16:23] src/van.cc:140: assign rank=9 to node role=worker, ip=10.80.40.12, port=50033, is_recovery=0
[07:16:23] src/van.cc:140: assign rank=8 to node role=server, ip=10.80.40.12, port=56485, is_recovery=0
[07:16:23] src/van.cc:140: assign rank=11 to node role=worker, ip=10.80.40.11, port=52213, is_recovery=0
[07:16:23] src/van.cc:478: ? => 9. Meta: request=0, timestamp=0, control={ cmd=ADD_NODE, node={ role=worker, id=9, ip=10.80.40.12, port=50033, is_recovery=0 role=server, id=8, ip=10.80.40.12, port=56485, is_recovery=0 role=worker, id=11, ip=10.80.40.11, port=52213, is_recovery=0 role=scheduler, id=1, ip=0.0.0.0, port=1234, is_recovery=0 } }. THIS IS NOT DATA MSG!
[07:16:23] src/van.cc:478: ? => 11. Meta: request=0, timestamp=1, control={ cmd=ADD_NODE, node={ role=worker, id=9, ip=10.80.40.12, port=50033, is_recovery=0 role=server, id=8, ip=10.80.40.12, port=56485, is_recovery=0 role=worker, id=11, ip=10.80.40.11, port=52213, is_recovery=0 role=scheduler, id=1, ip=0.0.0.0, port=1234, is_recovery=0 } }. THIS IS NOT DATA MSG!
[07:16:23] src/van.cc:478: ? => 8. Meta: request=0, timestamp=2, control={ cmd=ADD_NODE, node={ role=worker, id=9, ip=10.80.40.12, port=50033, is_recovery=0 role=server, id=8, ip=10.80.40.12, port=56485, is_recovery=0 role=worker, id=11, ip=10.80.40.11, port=52213, is_recovery=0 role=scheduler, id=1, ip=0.0.0.0, port=1234, is_recovery=0 } }. THIS IS NOT DATA MSG!
[07:16:23] src/van.cc:166: the scheduler is connected to 2 workers and 1 servers
[07:16:23] src/van.cc:503: 11 => 1. Meta: request=1, timestamp=1, control={ cmd=BARRIER, barrier_group=7 }. THIS IS NOT DATA MSG!
[07:16:23] src/van.cc:265: Barrier count for 7 : 1
[07:16:23] src/van.cc:478: ? => 1. Meta: request=1, timestamp=3, control={ cmd=BARRIER, barrier_group=7 }. THIS IS NOT DATA MSG!
[07:16:23] src/van.cc:503: 1 => 1. Meta: request=1, timestamp=3, control={ cmd=BARRIER, barrier_group=7 }. THIS IS NOT DATA MSG!
[07:16:23] src/van.cc:265: Barrier count for 7 : 2
```

logs on the server
```
[07:16:03] byteps/server/server.cc:339: BytePS server engine uses 4 threads, consider increasing BYTEPS_SERVER_ENGINE_THREAD for higher performance
[07:16:03] src/./zmq_van.h:61: BYTEPS_ZMQ_MAX_SOCKET set to 1024
[07:16:03] src/./zmq_van.h:66: BYTEPS_ZMQ_NTHREADS set to 4
[07:16:03] src/van.cc:389: Bind to role=server, ip=10.80.40.12, port=56485, is_recovery=0
[07:16:03] src/./zmq_van.h[:28707:16:03: Start ZMQ recv thread] src/van.cc:
478: ? => 1. Meta: request=0, timestamp=0, control={ cmd=ADD_NODE, node={ role=server, ip=10.80.40.12, port=56485, is_recovery=0 } }. THIS IS NOT DATA MSG!
[07:16:23] src/van.cc:503: 1 => 2147483647. Meta: request=0, timestamp=2, control={ cmd=ADD_NODE, node={ role=worker, id=9, ip=10.80.40.12, port=50033, is_recovery=0 role=server, id=8, ip=10.80.40.12, port=56485, is_recovery=0 role=worker, id=11, ip=10.80.40.11, port=52213, is_recovery=0 role=scheduler, id=1, ip=0.0.0.0, port=1234, is_recovery=0 } }. THIS IS NOT DATA MSG!
[07:16:23] src/van.cc:338: S[8] is connected to others
[07:16:23] src/van.cc:478: ? => 1. Meta: request=1, timestamp=1, control={ cmd=BARRIER, barrier_group=7 }. THIS IS NOT DATA MSG!
```

logs on worker 0
```
...
[2020-07-12 07:16:19.407214: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.1.bn2.weight, declared key (not PS key): 307 rank=2
[2020-07-12 07:16:19.407226: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.1.bn3.bias, declared key (not PS key): 308 rank=2
[2020-07-12 07:16:19.407236: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.1.bn3.weight, declared key (not PS key): 309 rank=2
[2020-07-12 07:16:19.407248: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.1.conv1.weight, declared key (not PS key): 310 rank=2
[2020-07-12 07:16:19.407258: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.1.conv2.weight, declared key (not PS key): 311 rank=2
[2020-07-12 07:16:19.407269: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.1.conv3.weight, declared key (not PS key): 312 rank=2
[2020-07-12 07:16:19.407282: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.2.bn1.bias, declared key (not PS key): 313 rank=2
[2020-07-12 07:16:19.407296: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.2.bn1.weight, declared key (not PS key): 314 rank=2
[2020-07-12 07:16:19.407306: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.2.bn2.bias, declared key (not PS key): 315 rank=2
[2020-07-12 07:16:19.407317: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.2.bn2.weight, declared key (not PS key): 316 rank=2
[2020-07-12 07:16:19.407328: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.2.bn3.bias, declared key (not PS key): 317 rank=2
[2020-07-12 07:16:19.407338: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.2.bn3.weight, declared key (not PS key): 318 rank=2
[2020-07-12 07:16:19.407350: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.2.conv1.weight, declared key (not PS key): 319 rank=2
[2020-07-12 07:16:19.407364: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.2.conv2.weight, declared key (not PS key): 320 rank=2
[2020-07-12 07:16:19.407376: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.2.conv3.weight, declared key (not PS key): 321 rank=2
[2020-07-12 07:16:19.408625: D byteps/common/operations.cc:245] byteps.Parameter.bn1.bias partitioned to 1 part(s), total_len=256, key_range=[10551296, 10551296] rank=2
[2020-07-12 07:16:19.408657: T byteps/common/operations.cc:259] Begin init byteps.Parameter.bn1.bias, size=256, parts=1
[2020-07-12 07:16:19.410433: T byteps/common/shared_memory.cc:44] initialized share memory size 4096
[2020-07-12 07:16:19.410469: T byteps/common/operations.cc:280] byteps.Parameter.bn1.bias: open shared memory size 256
[2020-07-12 07:16:19.410489: T byteps/common/operations.cc:311] Finish Init byteps.Parameter.bn1.bias, size=256, parts=1
[2020-07-12 07:16:19.410526: T byteps/common/operations.cc:197] EnqueueTensor: byteps.Parameter.bn1.bias_0, key=10551296, offset=0, len=256, device=2 rank=2
[2020-07-12 07:16:19.410547: T byteps/common/scheduled_queue.cc:94] Queue COORDINATE_REDUCE addTask: byteps.Parameter.bn1.bias_0 key: 10551296 rank: 2
[2020-07-12 07:16:19.410565: T byteps/common/operations.cc:211] EnqueueTensor finished: byteps.Parameter.bn1.bias, rank=2
[2020-07-12 07:16:19.410605: T byteps/common/scheduled_queue.cc:150] Queue COORDINATE_REDUCE getTask: byteps.Parameter.bn1.bias_0 key: 10551296 rank: 2
[2020-07-12 07:16:19.410650: T byteps/common/core_loops.cc:91] Rank=2 finishes COORDINATE_REDUCE, tensor: byteps.Parameter.bn1.bias_0, key=10551296; Passing to the next queue.
[2020-07-12 07:16:19.410663: T byteps/common/scheduled_queue.cc:94] Queue REDUCE addTask: byteps.Parameter.bn1.bias_0 key: 10551296 rank: 2
[2020-07-12 07:16:19.410706: T byteps/common/core_loops.cc:167] byteps.Parameter.bn1.bias_0 send coordinate info: Signal=0, rank=2, key=10551296
[2020-07-12 07:16:19.410728: T byteps/common/communicator.cc:199] root socket recved: src=2, signal=0, key=10551296, myrank=3
[07:16:23] src/van.cc:503: 1 => 2147483647. Meta: request=0, timestamp=1, control={ cmd=ADD_NODE, node={ role=worker, id=9, ip=10.80.40.12, port=50033, is_recovery=0 role=server, id=8, ip=10.80.40.12, port=56485, is_recovery=0 role=worker, id=11, ip=10.80.40.11, port=52213, is_recovery=0 role=scheduler, id=1, ip=0.0.0.0, port=1234, is_recovery=0 } }. THIS IS NOT DATA MSG!
[07:16:23] src/./zmq_van.h:287: Start ZMQ recv thread
[07:16:23] src/./zmq_van.h:287: Start ZMQ recv thread
[07:16:23] src/van.cc:338: W[11] is connected to others
[07:16:23] src/van.cc:478: ? => 1. Meta: request=1, timestamp=1, control={ cmd=BARRIER, barrier_group=7 }. THIS IS NOT DATA MSG!
```

logs on worker 1
```
[2020-07-12 07:16:23.558072: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.0.conv2.weight, declared key (not PS key): 299 rank=1
[2020-07-12 07:16:23.558084: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.0.conv3.weight, declared key (not PS key): 300 rank=1
[2020-07-12 07:16:23.558096: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.0.downsample.0.weight, declared key (not PS key): 301 rank=1
[2020-07-12 07:16:23.558108: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.0.downsample.1.bias, declared key (not PS key): 302 rank=1
[2020-07-12 07:16:23.558123: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.0.downsample.1.weight, declared key (not PS key): 303 rank=1
[2020-07-12 07:16:23.558136: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.1.bn1.bias, declared key (not PS key): 304 rank=1
[2020-07-12 07:16:23.558149: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.1.bn1.weight, declared key (not PS key): 305 rank=1
[2020-07-12 07:16:23.558160: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.1.bn2.bias, declared key (not PS key): 306 rank=1
[2020-07-12 07:16:23.558173: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.1.bn2.weight, declared key (not PS key): 307 rank=1
[2020-07-12 07:16:23.558185: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.1.bn3.bias, declared key (not PS key): 308 rank=1
[2020-07-12 07:16:23.558197: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.1.bn3.weight, declared key (not PS key): 309 rank=1
[2020-07-12 07:16:23.558210: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.1.conv1.weight, declared key (not PS key): 310 rank=1
[2020-07-12 07:16:23.558222: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.1.conv2.weight, declared key (not PS key): 311 rank=1
[2020-07-12 07:16:23.558233: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.1.conv3.weight, declared key (not PS key): 312 rank=1
[2020-07-12 07:16:23.558245: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.2.bn1.bias, declared key (not PS key): 313 rank=1
[2020-07-12 07:16:23.558258: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.2.bn1.weight, declared key (not PS key): 314 rank=1
[2020-07-12 07:16:23.558270: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.2.bn2.bias, declared key (not PS key): 315 rank=1
[2020-07-12 07:16:23.558281: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.2.bn2.weight, declared key (not PS key): 316 rank=1
[2020-07-12 07:16:23.558293: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.2.bn3.bias, declared key (not PS key): 317 rank=1
[2020-07-12 07:16:23.558306: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.2.bn3.weight, declared key (not PS key): 318 rank=1
[2020-07-12 07:16:23.558318: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.2.conv1.weight, declared key (not PS key): 319 rank=1
[2020-07-12 07:16:23.558330: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.2.conv2.weight, declared key (not PS key): 320 rank=1
[2020-07-12 07:16:23.558341: D byteps/common/global.cc:359] Declared tensor byteps.Parameter.layer4.2.conv3.weight, declared key (not PS key): 321 rank=1
[2020-07-12 07:16:23.559761: D byteps/common/operations.cc:245] byteps.Parameter.bn1.bias partitioned to 1 part(s), total_len=256, key_range=[10551296, 10551296] rank=1
[2020-07-12 07:16:23.559794: T byteps/common/operations.cc:259] Begin init byteps.Parameter.bn1.bias, size=256, parts=1
[2020-07-12 07:16:23.559807: D byteps/common/operations.cc:265] byteps.Parameter.bn1.bias is already on cpu, len=256
[2020-07-12 07:16:23.561610: T byteps/common/shared_memory.cc:44] initialized share memory size 4096
[2020-07-12 07:16:23.561636: T byteps/common/operations.cc:280] byteps.Parameter.bn1.bias: open shared memory size 256
[2020-07-12 07:16:23.561648: T byteps/common/operations.cc:311] Finish Init byteps.Parameter.bn1.bias, size=256, parts=1
[2020-07-12 07:16:23.561685: T byteps/common/operations.cc:197] EnqueueTensor: byteps.Parameter.bn1.bias_0, key=10551296, offset=0, len=256, device=-1 rank=1
[2020-07-12 07:16:23.561714: T byteps/common/scheduled_queue.cc:94] Queue COORDINATE_REDUCE addTask: byteps.Parameter.bn1.bias_0 key: 10551296 rank: 1
[2020-07-12 07:16:23.561733: T byteps/common/operations.cc:211] EnqueueTensor finished: byteps.Parameter.bn1.bias, rank=1
[2020-07-12 07:16:23.561765: T byteps/common/scheduled_queue.cc:150] Queue COORDINATE_REDUCE getTask: byteps.Parameter.bn1.bias_0 key: 10551296 rank: 1
[2020-07-12 07:16:23.561818: T byteps/common/core_loops.cc:91] Rank=5 finishes COORDINATE_REDUCE, tensor: byteps.Parameter.bn1.bias_0, key=10551296; Passing to the next queue.
[2020-07-12 07:16:23.561837: T byteps/common/scheduled_queue.cc:94] Queue REDUCE addTask: byteps.Parameter.bn1.bias_0 key: 10551296 rank: 1
[2020-07-12 07:16:23.561886: T byteps/common/core_loops.cc:167] byteps.Parameter.bn1.bias_0 send coordinate info: Signal=0, rank=1, key=10551296
[2020-07-12 07:16:23.561900: T byteps/common/communicator.cc:199] root socket recved: src=1, signal=0, key=10551296, myrank=3
[07:16:23] src/van.cc:503: 1 => 2147483647. Meta: request=0, timestamp=0, control={ cmd=ADD_NODE, node={ role=worker, id=9, ip=10.80.40.12, port=50033, is_recovery=0 role=server, id=8, ip=10.80.40.12, port=56485, is_recovery=0 role=worker, id=11, ip=10.80.40.11, port=52213, is_recovery=0 role=scheduler, id=1, ip=0.0.0.0, port=1234, is_recovery=0 } }. THIS IS NOT DATA MSG!
[07:16:23] src/./zmq_van.h:287: Start ZMQ recv thread
[07:16:23] src/./zmq_van.h:287: Start ZMQ recv thread
[07:16:23] src/van.cc:338: W[9] is connected to others
[07:16:23] src/van.cc:478: ? => 1. Meta: request=1, timestamp=1, control={ cmd=BARRIER, barrier_group=7 }. THIS IS NOT DATA MSG!
```

**Comment by bobzhuyb** [comment]
You are stuck in the NCCL stage within each machine before reaching the multi-machine communication stage.  Can you first verify whether single worker (with multiple GPUs) can work?

If you search `Queue REDUCE addTask` in your logs, how many can you see from different ranks?

**Comment by wangli1426** [comment]
Hi @bobzhuyb 

Thank you once again for your prompt response.

I just verified single worker training using exactly the same commands as provided in https://github.com/bytedance/byteps/blob/master/docs/step-by-step-tutorial.md#pytorch. The single node training run normally on both of the two workers.

I also search `Queue REDUCE addTask` in the logs.
on worker 0:
```
cat log.out |grep 'Queue REDUCE addTask'                                   
[2020-07-12 07:45:51.145230: T byteps/common/scheduled_queue.cc:94] Queue REDUCE addTask: byteps.Parameter.bn1.bias_0 key: 10551296 rank: 2
[2020-07-12 07:45:51.167350: T byteps/common/scheduled_queue.cc:94] Queue REDUCE addTask: byteps.Parameter.bn1.bias_0 key: 10551296 rank: 1
[2020-07-12 07:45:51.227026: T byteps/common/scheduled_queue.cc:94] Queue REDUCE addTask: byteps.Parameter.bn1.bias_0 key: 10551296 rank: 0
```
on worker 1:
```
cat log.out |grep 'Queue REDUCE addTask'
[2020-07-12 07:45:52. 96937: T byteps/common/scheduled_queue.cc:94] Queue REDUCE addTask: byteps.Parameter.bn1.bias_0 key: 10551296 rank: 0
[2020-07-12 07:45:52.187393: T byteps/common/scheduled_queue.cc:94] Queue REDUCE addTask: byteps.Parameter.bn1.bias_0 key: 10551296 rank: 1
[2020-07-12 07:45:52.216220: T byteps/common/scheduled_queue.cc:94] Queue REDUCE addTask: byteps.Parameter.bn1.bias_0 key: 10551296 rank: 2
```

**Comment by bobzhuyb** [comment]
Normally you should see four `Queue REDUCE addTask`. When you run it with a single worker, you can find four `Queue REDUCE addTask`, right?   Can you grep `rank=3` and let me see the log?

Also, can you set `DMLC_PS_ROOT_URI` to be an explicit IP on the scheduler, instead of `0.0.0.0`?

**Comment by wangli1426** [comment]
Thanks for you suggestion. 

I have now managed to run the benchmark after I made the following changes:
1. For the scheduler, I set `DMLC_PS_ROOT_URI` to the actual IP of the node;
2. For each of the node, I edited the `/etc/hosts` file and changed the IP of the hostname from `127.0.0.1` to the actual IP.

The problem can only be solved when I make the two changes at the same time. I tried to launch the benchmark when I only made one of the changes above and training process still stuck.
